Add RPC API reference - Closes #948 (#954)
LiskHQ/lisk-docs: modules/ROOT/pages/references/rpc-api.adoc

= WebSocket JSON 2.0 RPC API
:toc:

== Lisk Service WebSocket JSON-RPC API Documentation

=== Access paths and compatibility
The WebSocket API can be accessed by the `wss://service.lisk.io/rpc-v1`.

> You can also access the testnet network by `wss://testnet-service.lisk.io/rpc-v1`.

Reviewer: ```suggestion
The testnet network can also be accessed by `wss://testnet-service.lisk.io/rpc-v1`.
```

The Lisk Service WebSocket API uses the `socket.io` library and it is compatible with JSON-RPC 2.0 standard.
The specification below contains numerous examples how to use the API in practice.

=== Endpoint logic

> The logic of the endpoints comes as follows: the method naming is always based on the following pattern: `<action>.<entity>`, where the action is equivalent to HTTP standard (GET, POST, PUT, etc.) and `entity` is a part of the application logic, ex. `accounts`, transactions and so on.

Reviewer: ```suggestion
The logic of the endpoints is derived as follows: the method naming is always based on the following pattern: `<action>.<entity>`, where the action is equivalent to HTTP standard (GET, POST, PUT, etc.) and `entity` is a part of the application logic, ex. `accounts`, transactions and so on.
```

              "balance": "4448642000",
              "unconfirmedSignature": 0
            }
          ]
        },
        "multisignatureMemberships": [
          {
            "address": "4935562234363081651L",
            "balance": "4448642000",
            "lifetime": 48,
            "min": 2,
            "publicKey": "968ba2fa993ea9dc27ed740da0daf49eddd740dbd7cb1cb4fc5db3a20baf341b",
            "secondPublicKey": "968ba2fa993ea9dc27ed740da0daf49eddd740dbd7cb1cb4fc5db3a20baf341b",
          }
        ],
        "transactionCount": {
          "incoming": "216",
          "outgoing": "1581"
        },
        "unconfirmedMultisignatureMemberships": [
          {
            "address": "4935562234363081651L",
            "publicKey": "968ba2fa993ea9dc27ed740da0daf49eddd740dbd7cb1cb4fc5db3a20baf341b",
            "secondPublicKey": "968ba2fa993ea9dc27ed740da0daf49eddd740dbd7cb1cb4fc5db3a20baf341b",
            "balance": "4448642000",
            "unconfirmedSignature": 0
          }
        ]
      }
    ],
    "meta": {
      "count": 100,
      "offset": 25,
      "total": 43749
    }
  }
]
----

> .Example: Get address with certain Lisk account ID

Reviewer: ```suggestion
.Example: Get address with a specific Lisk account ID
```

Author: ```suggestion
.Example: Get account with a specific Lisk account ID
```

[source,json]
----
{
    "method": "get.accounts",
    "params": {
        "address": "123L"
    }
}
----

> ==== get.votes

Reviewer 2: `get.votes_sent`

Update: we've changed it to make it less ambiguous, the old API call `get.votes` still work as legacy API.

Retrieves votes of a single account based on address, public key or delegate name.

Supports pagination.

[cols="2,3,4", options="header"]
.Parameters
|===
|Parameter name |Example |Description

|`address`
|`4935562234363081651L`
|Lisk account address

|`publickey`
|`968ba2...af341b`
|Public key to query

|`secpubkey`
|`968ba2...af341b`
|Second public key to query

|`username`
|`genesis_12`
|Delegate username to query

|`limit`
|`0`
|Number of requested items

|`offset`
|`25`
|Number of items to skip

|===

.Response
[source,json]
----
{
  "data": [
    {
      "address": "4935562234363081651L",
      "publicKey": "968ba2fa993ea9dc27ed740da0daf49eddd740dbd7cb1cb4fc5db3a20baf341b",
      "balance": 1081560729258,
      "username": "liskhq"
    }
  ],
  "meta": {
    "count": 100,
    "offset": 25,
    "total": 43749
  }
}
----

.Example
[source,json]
----
{
    "method": "get.votes",
    "params": {
        "address": "4935562234363081651L"
    }
}
----

> ==== get.voters

Reviewer 2: `get.votes_received`

      "rank": 93,
      "rewards": "109500000000",
      "username": "genesis_84",
      "vote": "4630668157412954"
    }
  ],
  "meta": {
    "count": 100,
    "offset": 25,
    "total": 43749
  }
}
----
.Example
[source,json]
----
{
    "method": "get.delegates",
    "params": {
        "address": "4935562234363081651L"
    }
}
----
==== get.delegates.next_forgers
Retrieves next forgers with details in the current round.

Supports pagination.

[cols="2,3,4", options="header"]
.Parameters
|===
|Parameter name |Example |Description

|`limit`
|`50`
|Limit applied to results

|`offset`
|`25`
|Number of items to skip

|===

.Response
[source,json]
----
{
  "data": [
    {
      "address": "4935562234363081651L",
      "approval": "35.77",
      "missedBlocks": 157,
      "producedBlocks": 55222,
      "productivity": "99.72",
      "publicKey": "6122ac1fd71b437014ddbc4ec01e07879f5af1853536efaa0233bc12907c684b",
      "secondPublicKey": "6122ac1fd71b437014ddbc4ec01e07879f5af1853536efaa0233bc12907c684b",
      "rank": 93,
      "username": "genesis_84",
      "vote": "4630668157412954"
    }
  ],
  "meta": {
    "count": 100,
    "offset": 25,
    "total": 43749
  },
}
----

> .Example: Get 20 item, skip 50 first

Reviewer 2: "Get 20 items, ..."

|`httpPort`
|`8000`
|HTTP port of the node or delegate

|`wsPort	`
|`8001`
|Web socket port for the node or delegate

|`os`
|`debian`
|OS of the node

|`version`
|`v0.8.0`
|Lisk version of the node

|`state`
|`2`
|Current state of the network

|`height`
|`8350681`
|Current height of the network

|`broadhash`
|`258974...3db1ea`
|Broadhash of the network

|`limit`
|`0`
|Number of requested items

|`offset`
|`25`
|Number of items to skip

|`sort`
|`name:desc`
|Sorting order

Available values: `height:asc`, `height:desc`, `version:asc`, `version:desc`

Default value: `height:desc`

|===

.Response
[source,json]
----
{
  "data": [
    {
      "ip": "127.0.0.1",
      "httpPort": 8000,
      "wsPort": 8001,
      "os": "debian",
      "version": "v0.8.0",
      "state": v0.8.0,
      "height": 8350681,
      "broadhash": "258974416d58533227c6a3da1b6333f0541b06c65b41e45cf31926847a3db1ea",
      "nonce": "sYHEDBKcScaAAAYg",
      "location": {
        "city": "Berlin",
        "countryCode": "DE",
        "countryName": "Germany",
        "hostname": "host.210.239.23.62.rev.coltfrance.com",
      }
    }
  ],
  "meta": {
    "count": 100,
    "offset": 25,
    "total": 43749
  }
}
----

> .Example: Get hosts with certain IP

Reviewer: ```suggestion
.Example: Get hosts with a specific IP address
```

|`25`
|Number of items to skip

|===

.Response
[source,json]
----
{
  "data":{
    "timeline":[
      {
        "date":"2019-12",
        "transactionCount":503,
        "volume":"20087361290583",
        "timestamp":1574812800
      },
      {
        "date":"2019-11",
        "transactionCount":363,
        "volume":"61577610129315",
        "timestamp":1574726400
      },
      {
        "date":"2019-10",
        "transactionCount":1211,
        "volume":"27747743232112",
        "timestamp":1574640000
      },
    ],
    "distributionByType":{
      "0":7063,
      "1":1,
      "2":2,
      "3":51
    },
    "distributionByAmount":{
      "1_10":5042,
      "10_100":585,
      "100_1000":542,
      "1000_10000":236,
      "10000_100000":52,
      "100000_1000000":3,
      "0.1_1":657
    }
  },
  "meta":{
    "limit":3,
    "offset":0,
    "aggregateBy":"month",
    "dateFormat":"YYYY-MM",
    "dateFrom":"2019-10",
    "dateTo":"2019-12"
  },
  "links":{

  }
}
----
.Example: Get transaction statistics for past 12 months
[source,json]
----
{
    "method": "get.transactions.statistics.month",
    "params": {
        "limit": 12,
    }
}
----
=== Network
==== get.network.status

> Retrieves network details and constants such as network height, broadhash, fees, reward amount etc.

Reviewer: ```suggestion
Retrieves network details and constants such as network height, broadhash, fees, reward amount, etc.
```

No parameters.

.Response
[source,json]
----
{
  "broadhash": "258974416d58533227c6a3da1b6333f0541b06c65b41e45cf31926847a3db1ea",
  "height": 123,
  "networkHeight": 123,
  "epoch": "2016-05-24T17:00:00.000Z",
  "milestone": "500000000",
  "nethash": "ed14889723f24ecc54871d058d98ce91ff2f973192075c0155ba2b7b70ad2511",
  "supply": "10575384500000000",
  "reward": "500000000",
  "fees": {
    "send": "10000000",
    "vote": "100000000",
    "secondSignature": "500000000",
    "delegate": "2500000000",
    "multisignature": "500000000",
    "dappRegistration": "2500000000",
    "dappWithdrawal": "10000000",
    "dappDeposit": "10000000"
  }
}
----
.Example
[source,json]
----
{
    "method": "get.network.status"
}
----

> ==== get.network.statistic

Reviewer 2: `get.network.statistics`

> Retrieves network statistics such as number of peers, node versions, heights etc.

Reviewer: ```suggestion
Retrieves network statistics such as the number of peers, node versions, heights, etc.
```
